controllers/baremetalset_controller.go

```diff
@@ -90,76 +102,139 @@ func (r *BaremetalSetReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
-	baremetalHostsList := &metal3v1alpha1.BareMetalHostList{}
-	listOpts := []client.ListOption{
-		client.InNamespace("openshift-machine-api"),
-	}
-	err = r.Client.List(context.TODO(), baremetalHostsList, listOpts...)
-	if err != nil {
+	// First we need the public SSH key, which should be stored in a secret
+	hashes := []ospdirectorv1beta1.Hash{}
+	sshSecret, secretHash, err := common.GetSecret(r, instance.Spec.DeploymentSSHSecret, instance.Namespace)
+	if err != nil && errors.IsNotFound(err) {
+		return ctrl.Result{RequeueAfter: time.Second * 20}, fmt.Errorf("DeploymentSSHSecret secret does not exist: %v", err)
+	} else if err != nil {
 		return ctrl.Result{}, err
 	}
 
-	for _, baremetalHost := range baremetalHostsList.Items {
+	// TODO: Put hashes in BaremetalSet status?
+	hashes = append(hashes, ospdirectorv1beta1.Hash{Name: sshSecret.Name, Hash: secretHash})
 
-		if baremetalHost.Spec.ConsumerRef != nil || baremetalHost.Spec.Online == true {
-			continue
-		}
-		r.Log.Info(fmt.Sprintf("Available BaremetalHost: %s", baremetalHost.ObjectMeta.Name))
-		availableBaremetalHosts = append(availableBaremetalHosts, baremetalHost.ObjectMeta.Name)
+	// If BaremetalHosts status map is nil, create it
+	if instance.Status.BaremetalHosts == nil {
+		instance.Status.BaremetalHosts = map[string]ospdirectorv1beta1.BaremetalHostStatus{}
 	}
 
-	// for each available BaremetalHost we Update the reference to use our image
-	for _, availableHost := range availableBaremetalHosts {
-
-		foundBaremetalHost := &metal3v1alpha1.BareMetalHost{}
-		err = r.Client.Get(context.TODO(), types.NamespacedName{Name: availableHost, Namespace: "openshift-machine-api"}, foundBaremetalHost)
-
-		if err != nil {
-			return ctrl.Result{}, err
-		}
+	// Provision / deprovision requested replicas
+	if err := r.ensureBaremetalHosts(instance, provisionServer, sshSecret); err != nil {
+		return ctrl.Result{}, err
+	}
 
-		r.Log.Info(fmt.Sprintf("Updating BaremetalHost: %s", availableHost))
-		foundBaremetalHost.Spec.Online = true
-		foundBaremetalHost.Spec.ConsumerRef = &corev1.ObjectReference{Name: baremetalset.Name, Kind: "BaremetalSet"}
-		foundBaremetalHost.Spec.Image = &metal3v1alpha1.Image{
-			URL:      provisionServer.Status.LocalImageURL,
-			Checksum: fmt.Sprintf("%s.md5sum", provisionServer.Status.LocalImageURL),
-		}
-		err = r.Client.Update(context.TODO(), foundBaremetalHost)
-		if err != nil {
-			return ctrl.Result{}, err
-		}
-		return ctrl.Result{RequeueAfter: time.Second * 5}, err
+	err = r.Client.Status().Update(context.TODO(), instance)
 
+	if err != nil {
+		r.Log.Error(err, "Failed to update CR status %v")
+		return ctrl.Result{}, err
 	}
 
 	return ctrl.Result{}, nil
 }
 
 // SetupWithManager - prepare controller for use with operator manager
 func (r *BaremetalSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	openshiftMachineApiSecretsFn := handler.ToRequestsFunc(func(o handler.MapObject) []reconcile.Request {
+		cc := mgr.GetClient()
```

Comment:
we don't have to get the client r.Client can be used

Reply:
Ack